### ISSUE TITLE
Accute accent character crashes SASPy

### ISSUE THREAD

**Issue by RoyalTS** [sassoftware/saspy#207]
**Describe the bug**

Trying to transfer a dataframe containing the accute accent character crashes the SASPy session.

**To Reproduce**

```python
import saspy
import pandas as pd
sas = saspy.SASsession(...)
df = pd.DataFrame({'name': ['´']})
sas.df2sd(df, table='test', libref='work')
```

yields

```
ValueError: invalid literal for int() with base 10: 'No SAS process attached. SAS process has terminated unexpectedly.'
```

**Desktop (please complete the following information):**
* OS: Linux
* SAS Version: 7.1.3
* SASPy Version: 2.4.1

**Comment by tomweber-sas** [comment]
Well, that's no good. Let's see what encodings you're running with.
Can you submit your SASsession object:
```
import saspy
import pandas as pd
sas = saspy.SASsession(...)
sas
```
That way I can see the SAS session encoding, what you told me the SAS session encoding was, what access method, ...

Thanks!
Tom

**Comment by RoyalTS** [comment]
```
Access Method         = IOM
SAS Config name       = iomlinux
WORK Path             = WITHHELD
SAS Version           = 9.04.01M4P11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = latin9
Python Encoding value = latin1
SAS process Pid value = 15279
```

**Comment by tomweber-sas** [comment]
ok, so SAS is running with Latin9, but saspy was told it was running Latin1. 
Can you change the encoding value in your configuration definition (iomlinux) to be latin9 and see if that solves it?
I'm going to try to reproduce this as well, but I've already see it work fine with other encodings (hadn't tried latin9 yet) for both STDIO and IOM when the encodings match. 
Tom

**Comment by tomweber-sas** [comment]
Well, I've reproduced what you reported. It is actually the SAS session itself that abends when I run this.

Here's a bit of that
```
NOTE: 20190207:10.39.31.19: 00000011:NOTE: New client connection (2) accepted from server port 28591 for user sas. Encryption level
      is Credentials using encryption algorithm SASPROPRIETARY. Peer IP address and port are [::ffff:127.0.0.1]:58504 for
      APPNAME=SASPy.
20190207:10.40.04.60: 00000004:5
                    The SAS System                                                                                         10:39
Thursday, February 7, 2019
20190207:10.40.04.60: 00000004:
20190207:10.40.04.60: 00000004:
20190207:10.40.04.60: 00000004:ERROR:  An exception has been encountered.
20190207:10.40.04.60: 00000004:ERROR: Please contact technical support and provide them with the following traceback information:


2                                                          The SAS System                           10:30 Thursday, February 7, 2019

20190207:10.40.04.60: 00000004:ERROR:
20190207:10.40.04.60: 00000004:ERROR: The SAS task name is [DATASTEP]
20190207:10.40.04.60: 00000004:ERROR: Segmentation Violation
20190207:10.40.04.60: 00000004:ERROR:
20190207:10.40.04.60: 00000004:ERROR: Traceback of the Exception:
20190207:10.40.04.60: 00000004:ERROR:
```

When I set the saspy encoding value to match SAS, then I get a different error - when using Latin9 - it works fine with latin1 or windows latin1.
The problem is that that character isn't in the Latin9 code page. So you get a python error that it can't transcode it, which is correct. The method just fails, and SAS stays up running, and it's just an error because you can't do that.

Here's a sample just using python like what is happening under the covers in df2sd(), both cases get the same error:

```
>>> sas
Access Method         = IOM
SAS Config name       = iomj
WORK Path             = /sastmp/SAS_workDDB100003F10_tom64-3/SAS_workB0C300003F10_tom64-3/
SAS Version           = 9.04.01M4P11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = latin9
Python Encoding value = latin9
SAS process Pid value = 16144


>>> df = pd.DataFrame({'name': ['´']})
>>> df
  name
0    ´
>>>
>>> sas.df2sd(df, table='test')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tom/github/saspy/saspy/sasbase.py", line 742, in df2sd
    return self.dataframe2sasdata(df, table, libref, results, keep_outer_quotes)
  File "/opt/tom/github/saspy/saspy/sasbase.py", line 767, in dataframe2sasdata
    self._io.dataframe2sasdata(df, table, libref, keep_outer_quotes)
  File "/opt/tom/github/saspy/saspy/sasioiom.py", line 1131, in dataframe2sasdata
    col_l = df[df.columns[name]].astype(str).apply(self._getbytelen).max()
  File "/usr/lib64/python3.5/site-packages/pandas/core/series.py", line 2510, in apply
    mapped = lib.map_infer(values, f, convert=convert_dtype)
  File "pandas/_libs/src/inference.pyx", line 1521, in pandas._libs.lib.map_infer
  File "/opt/tom/github/saspy/saspy/sasioiom.py", line 1110, in _getbytelen
    return len(x.encode(self.sascfg.encoding))
  File "/usr/lib64/python3.5/encodings/iso8859_15.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_table)
UnicodeEncodeError: 'charmap' codec can't encode character '\xb4' in position 0: character maps to <undefined>
>>>
>>>
>>>
>>> x = '´'
>>> x
'´'
>>> x.encode()
b'\xc2\xb4'
>>> x.encode('latin1')
b'\xb4'
>>> x.encode('latin9')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.5/encodings/iso8859_15.py", line 12, in encode
    return codecs.charmap_encode(input,errors,encoding_table)
UnicodeEncodeError: 'charmap' codec can't encode character '\xb4' in position 0: character maps to <undefined>
>>>
```
As you can see, it's that that character doesn't exist in latin9 to be able to transcode it. 

Do you see the same behavior? 

Thanks,
Tom

**Comment by RoyalTS** [comment]
That makes sense. However, even after setting the right encoding this still crashes the entire connection rather than throwing UnicodeEncodeError for me.

**Comment by tomweber-sas** [comment]
Ah, I believe that is going to be due to a fix I have at master (the code I'm running) that I expect you don't have yet. That fix is https://github.com/sassoftware/saspy/commit/1c42ba4f43c8bfea8b0d53d135c58806c93055f2

That's where I get the python conversion error. This fix, was to get the correct server side byte length for string columns in the data frame. It also will happen to trap this situation and fail on the client side, not taking down IOM. Can you try again from master and see if it behaves the same for you with that?

I've also been debugging this some this morning. I do see that it's the IOM Server side (in the SAS session) component which it abending in this situation (Segmentation Violation; Null pointer for a GCB, that clearly isn't expected to be able to happen). I haven't tracked it to its origin, but I expect an unhandled transcoding exception that then sends things off the rails. That then causeses the IOM Client side (the Java process I've created from saspy) to terminate, thus, the message about 
 'No SAS process attached. SAS process has terminated unexpectedly.'

I can't fix IOM, so I am looking to see what I can do in saspy to mitigate this. Catching the transcode error (as I am at master) is part of this, I don't send this to IOM in that case, so it won't crash.
The other thing I'm looking at is to get, from out NLS folks, the mapping of all of the SAS session encodings to the Python equivalent, so that I can set the 'encoding' value in saspy myself based upon the SAS session encoding, so that the original problem wont happen to begin with; the encodings don't match. 

Those are the only two things I can really do in saspy to avoid this and help eliminate the configuration mistake. Do you have any other thoughts on that, which I may be missin, that could help? Can you confirm the behavior I see when you run from what's at master?

Thanks!
Tom

**Comment by tomweber-sas** [comment]
I've received the mappings between the SAS session encodings and the Python Codecs (https://docs.python.org/3.5/library/codecs.html#standard-encodings) from our NLS team.
I've integrated that into saspy such that it should no longer be necessary to supply the encoding value in the configuration definition. Of course, you still can, and I will validate it to be sure it's correct; no changes to any existing user code necessary.

Here's a sample of cases where I specify different encodings (overriding what's in the config definition).
```
 tom64-3> python3.5
Python 3.5.5 (default, Feb  6 2018, 10:56:47)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-18)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>>
>>> sas = saspy.SASsession(cfgname='sdssas')
SAS Connection established. Subprocess id is 8418

>>> sas
Access Method         = STDIO
SAS Config name       = sdssas
WORK Path             = /sastmp/SAS_workB1D8000020FE_tom64-3/
SAS Version           = 9.04.01M4D11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 8446


>>> sas1 = saspy.SASsession(cfgname='sdssas', encoding='')
SAS Connection established. Subprocess id is 8463

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to cp1252 based upon the SAS session encoding value of wlatin1.
>>>
>>> sas1
Access Method         = STDIO
SAS Config name       = sdssas
WORK Path             = /sastmp/SAS_workA59B0000212B_tom64-3/
SAS Version           = 9.04.01M4D11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 8491


>>> sas2 = saspy.SASsession(cfgname='sdssas', encoding='euc_cn')
SAS Connection established. Subprocess id is 8507

The encoding value provided doesn't match the SAS session encoding.
SAS encoding is wlatin1. Specified encoding is euc_cn.
Using encoding cp1252 instead to avoid transcoding problems.
You can override this change, if you think you must, by changing the encoding attribute of the SASsession object, as follows.
If you had 'sas = saspy.SASsession(), then submit: "sas._io.sascfg.encoding='override_encoding'" to change it
>>>
>>> sas2
Access Method         = STDIO
SAS Config name       = sdssas
WORK Path             = /sastmp/SAS_work759000002157_tom64-3/
SAS Version           = 9.04.01M4D11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 8535


>>> sas3 = saspy.SASsession(cfgname='sdssas', encoding='euwc_cn')
The encoding provided (euwc_cn) doesn't exist in this Python session. Setting it to ''.
The correct encoding will attempt to be determined based upon the SAS session encoding.
SAS Connection established. Subprocess id is 8551

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to cp1252 based upon the SAS session encoding value of wlatin1.
>>>
>>> sas3
Access Method         = STDIO
SAS Config name       = sdssas
WORK Path             = /sastmp/SAS_work29B100002183_tom64-3/
SAS Version           = 9.04.01M4D11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 8579


>>>
SAS Connection terminated. Subprocess id was 8463
SAS Connection terminated. Subprocess id was 8507
SAS Connection terminated. Subprocess id was 8551
SAS Connection terminated. Subprocess id was 8418
tom64-3>
```

Also, for the case where no encoding is provided, and the SAS_config_options 'verbose' is turned off, you won't get the messages about it not being set and what it got set to (that would be considered verbose 'noise' for the expected normal use case); sas1 in the samples above:
```
>>> saspy.sascfg_personal.SAS_config_options['verbose']
True
>>> saspy.sascfg_personal.SAS_config_options['verbose']=False
>>> saspy.sascfg_personal.SAS_config_options['verbose']
False
>>> sas1 = saspy.SASsession(cfgname='sdssas', encoding='')
>>>
>>> sas1
Access Method         = STDIO
SAS Config name       = sdssas
WORK Path             = /sastmp/SAS_workB352000021E5_tom64-3/
SAS Version           = 9.04.01M4D11092016
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 8677


>>>
```
I haven't pushed this to master yet. I'll post when I do. Thought I'd see if you had any feedback on it while I'm finishing up validating all of it.

Thanks,
Tom

**Comment by tomweber-sas** [comment]
@RoyalTS I've pushed this, along with other enhancements at master, and I've built a new release and pypi; V2.4.2. You should no longer need to specify the encoding yourself and you also shouldn't have IOM failues due to transcoding problems for df2sd. Let me know if you're ok to close this, or if you still have any outstanding questions or issue with this.

Thanks!
Tom

**Comment by tomweber-sas** [comment]
I just got back to looking at this and I could add an option to force the data to go through, replacing any invalid characters via pythons encode(encoding ,errors='replace'). This wouldn't be the default, as it adds the extra processing of all that transcoding, but it would allow the data to go through and get to SAS without IOM crashing. The alternative is to have to clean up those characters in the data frame yourself before using df2sd(). Do you think this would be useful, or is cleaning the data on the python side and then sending it to SAS what you would do instead? The python replacement character is '?', BTW.

Here's an example of what I'm thinking:
```
tom64-3> python3.5
Python 3.5.5 (default, Feb  6 2018, 10:56:47)
[GCC 4.4.7 20120313 (Red Hat 4.4.7-18)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession(cfgname='sdssas')
SAS Connection established. Subprocess id is 29160

No encoding value provided. Will try to determine the correct encoding.
Setting encoding to iso8859_15 based upon the SAS session encoding value of latin9.
>>>
>>> cars = sas.sasdata('cars', 'sashelp')
>>> dfc = cars.to_df()
>>> dfc.iloc[425,0]='´'
>>> dfc.tail()
      Make                    Model   Type  Origin DriveTrain   MSRP  Invoice  \
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  40565    38203
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  42565    40083
425      ´               S80 T6 4dr  Sedan  Europe      Front  45210    42573
426  Volvo                      V40  Wagon  Europe      Front  26135    24641
427  Volvo                     XC70  Wagon  Europe        All  35145    33112

     EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
423         2.4        5.0         197        21           28    3450
424         2.3        5.0         242        20           26    3450
425         2.9        6.0         268        19           26    3653
426         1.9        4.0         170        22           29    2822
427         2.5        5.0         208        20           27    3823

     Wheelbase  Length
423        105     186
424        105     186
425        110     190
426        101     180
427        109     186
>>>
>>> x = sas.df2sd(dfc)
Transcoding failure. Replacing invalid characters with '?'. Column 0 of row:
Pandas(Make='´', Model='S80 T6 4dr', Type='Sedan', Origin='Europe', DriveTrain='Front', MSRP=45210, Invoice=42573, EngineSize=2.9, Cylinders=6.0, Horsepower=268, MPG_City=19, MPG_Highway=26, Weight=3653, Wheelbase=110, Length=190)
>>> x.tail()
    Make                    Model   Type  Origin DriveTrain   MSRP  Invoice  \
0  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  40565    38203
1  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  42565    40083
2      ?               S80 T6 4dr  Sedan  Europe      Front  45210    42573
3  Volvo                      V40  Wagon  Europe      Front  26135    24641
4  Volvo                     XC70  Wagon  Europe        All  35145    33112

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         2.4          5         197        21           28    3450
1         2.3          5         242        20           26    3450
2         2.9          6         268        19           26    3653
3         1.9          4         170        22           29    2822
4         2.5          5         208        20           27    3823

   Wheelbase  Length
0        105     186
1        105     186
2        110     190
3        101     180
4        109     186
>>>
```
Any thoughts?

Thanks,
Tom

**Comment by RoyalTS** [comment]
I wouldn't clean up the data by default. That'll just use up time unnecessarily in most cases where the encoding is fine. Would leave it up to the user to deal with this, which isn't much of a problem _as long as the way this problem manifests isn't the session crashing completely_, which it now no longer does. So, fine as it is now, as far as I'm concerned.

**Comment by tomweber-sas** [comment]
Oh no, never by default. That's data corruption. I appreciate the feedback. I think I will leave it as is, other than catching the transcoding exception and print out a nicer error message. The user can simply do what I would (if I gave them a manual override to force it through), just encode it w/ errors='replace' (using the SAS session encoding) then decode it back, and it will be utf-8 which will then convert to the SAS session encoding w/out issue because the invalid chars will now be the valid replacement chars.
I've pushed the change to catch it and print a message to master.
Suppose we can close this one then?

Thanks!
Tom
